### ISSUE TITLE
Regressor Chain at ML

### ISSUE THREAD

**Issue by sangmandu** [sangmandu/Optimus#9]
## 결론
머신러닝에서는 큰 효과가 없었습니다. 기존 베이스라인 코드의 성능과 크게 변화가 없습니다. 딥러닝에서 다시 한번 진행해봐야 겠습니다.

- [딥러닝 구현 시 참고 블로그](https://data-newbie.tistory.com/845)

## 코드
```py
df = pd.read_csv('./train.csv')

train_df, valid_df = train_test_split(df, train_size=0.8)

train_x = train_df.filter(regex='X') # Input : X Featrue
train_y = train_df.filter(regex='Y') # Output : Y Feature

valid_x = valid_df.filter(regex='X') # Input : X Featrue
valid_y = valid_df.filter(regex='Y') # Output : Y Feature

test_x = pd.read_csv('./test.csv').drop(columns=['ID'])

LR = MultiOutputRegressor(LinearRegression()).fit(StandardScaler().fit_transform(train_x), train_y)

valid_preds = LR.predict(StandardScaler().fit_transform(valid_x))
lg_nrmse(valid_y.values, valid_preds)
```
```
(1.9794107500192726,
 [0.2570976462275067,
  0.3564070925110959,
  0.3489148464672859,
  0.19128313311864284,
  0.08052908347240303,
  0.11776446114654164,
  0.12998308967136904,
  0.024563404824937202,
  0.024177890370408312,
  0.04045698821726977,
  0.033867467032633485,
  0.024399374844050643,
  0.024300513388101932,
  0.02435720723906979])
```

```py
pd.set_option('display.max_columns', None)

cor = train_x.drop(columns=['X_04', 'X_23', 'X_47', 'X_48']).corr(method='pearson')

for col, rows in cor.iterrows():
    rows = [(row, idx) for idx, row in enumerate(rows, 1)]
    print(col, " : ", sorted(rows, reverse=True))
```
```
X_01  :  [(1.0, 1), (0.626033203977422, 5), (0.553404509444649, 4), (0.3240111425404451, 21), (0.2970008586459092, 18), (0.2776905707266618, 19), (0.2736756726988027, 20), (0.2168517059291016, 14), (0.21054714813766032, 3), (0.20833871643958984, 16), (0.1691913486054928, 38), (0.14381006678294225, 13), (0.12336251694229391, 15), (0.11900845513829597, 28), (0.11772936986684726, 17), (0.11683669742465277, 36), (0.0994738615461247, 30), (0.09117659459411666, 42), (0.07303576196583492, 37), (0.06629064792520797, 40), (0.04772855663664555, 49), (0.016113686366053023, 45), (0.014397842757909653, 32), (0.014090033663099534, 48), (0.012841848599199516, 47), (0.007393687353210759, 7), (0.003998961322438305, 52), (0.0037643650911230084, 44), (0.003214280182009499, 9), (0.0027372730270618303, 10), (0.0015731482828996858, 50), (6.770454852481871e-05, 12), (-0.0014355837454576928, 34), (-0.003999234933335353, 35), (-0.009222578042278851, 33), (-0.010286035476606368, 6), (-0.03361906458734867, 8), (-0.04022441040503725, 51), (-0.04167636900770578, 46), (-0.057777806221807206, 39), (-0.0811566414761905, 22), (-0.08231767100310527, 29), (-0.0829255394635855, 24), (-0.08632023032490588, 2), (-0.10240376934557106, 31), (-0.10782631484985758, 23), (-0.11315236937497056, 41), (-0.12056199479611536, 43), (-0.12271178193394831, 27), (-0.12553171874705013, 26), (-0.13224072513796778, 11), (-0.14261606167509092, 25)]
X_02  :  [(1.0, 2), (0.02509162765430731, 43), (0.02317605497832181, 31), (0.015628723542347366, 27), (0.013688031657865155, 41), (0.01340829833070978, 25), (0.012554515011376426, 29), (0.010487609403053343, 26), (0.008897144425079261, 24), (0.007041626564874387, 33), (0.005895358928213765, 23), (0.0028531552244811706, 35), (0.002797060763621059, 22), (0.0026306457902501037, 8), (0.0023059609328606487, 39), (0.002280590142417283, 46), (0.0015913596179644477, 11), (0.0012362036337449285, 9), (0.00042520860910321394, 7), (-6.434441508617248e-05, 10), (-0.0018950497493852602, 30), (-0.0024081217121033898, 51), (-0.004152625137197657, 34), (-0.00444759263039353, 45), (-0.004498766771855619, 48), (-0.004747667748327881, 50), (-0.005033841136461374, 17), (-0.005275816945939184, 47), (-0.005717453920057395, 32), (-0.006736209855836604, 28), (-0.006998368156048842, 6), (-0.006998398050726751, 13), (-0.007674154702297182, 52), (-0.009467283286097562, 44), (-0.012437140567370391, 12), (-0.01288005042812345, 49), (-0.014116307993429288, 15), (-0.014751380495833754, 37), (-0.015237834284837284, 14), (-0.017009271975266433, 38), (-0.02006705367792871, 36), (-0.021416801094031837, 42), (-0.02164680962660049, 40), (-0.025033239757870574, 16), (-0.027597903801821925, 3), (-0.03375199859710995, 20), (-0.03826841854648534, 18), (-0.0383707096420803, 19), (-0.04580515405532247, 21), (-0.045861720263782116, 5), (-0.04752171605122652, 4), (-0.08632023032490588, 1)]
```
- 출력 크기 상 X2까지만 작성합니다.

```py
pd.set_option('display.max_columns', None)

cor = train_y.corr(method='pearson')

for col, rows in cor.iterrows():
    rows = [(row, idx) for idx, row in enumerate(rows, 1)]
    print(col, " : ", sorted(rows, reverse=True))
```
```
Y_01  :  [(1.0, 1), (0.8742956572813272, 3), (0.7797158080651537, 2), (0.3929449218000198, 10), (0.3317785192505406, 11), (0.30267493523117395, 6), (0.2516040629755018, 8), (0.2508361285323926, 14), (0.2486376120681256, 13), (0.24813061814812823, 12), (0.2461552146334526, 9), (0.22809981206083352, 5), (0.15371660502404008, 4), (-0.14428361791636357, 7)]
Y_02  :  [(1.0, 2), (0.8463252977555727, 3), (0.7797158080651537, 1), (0.36292380102783184, 10), (0.30691324800315334, 11), (0.25749435938619814, 8), (0.25729244159981884, 14), (0.2566803013931644, 13), (0.25514608687807955, 12), (0.25372591683268725, 6), (0.25299286812530203, 9), (0.16913010693648214, 5), (0.056493786864361455, 4), (-0.17005297451014523, 7)]
Y_03  :  [(1.0, 3), (0.8742956572813272, 1), (0.8463252977555727, 2), (0.24675491776253486, 10), (0.1915809782296365, 6), (0.15880599536688572, 11), (0.10115230037731582, 5), (0.09668211540353398, 14), (0.09537062250924866, 13), (0.09505592667623942, 8), (0.09374294416980566, 12), (0.09168241279750015, 9), (-0.010105559419065498, 4), (-0.03732787909738924, 7)]
Y_04  :  [(1.0, 4), (0.7037047933817157, 5), (0.6206517173616585, 11), (0.5419338116516438, 10), (0.5404804991614621, 14), (0.5399477570597936, 8), (0.539042778266852, 12), (0.5371589859920781, 13), (0.533946795654652, 9), (0.25626178061170546, 6), (0.15371660502404008, 1), (0.056493786864361455, 2), (-0.010105559419065498, 3), (-0.556018724032774, 7)]
Y_05  :  [(1.0, 5), (0.7037047933817157, 4), (0.6887736652003423, 11), (0.542681269994975, 14), (0.5397582616820021, 8), (0.5386451410104928, 12), (0.5366804810401934, 13), (0.5350426077994276, 9), (0.5299272998756622, 10), (0.22809981206083352, 1), (0.2026368890085653, 6), (0.16913010693648214, 2), (0.10115230037731582, 3), (-0.601411890692791, 7)]
Y_06  :  [(1.0, 6), (0.6981263213927064, 10), (0.41491013123121523, 12), (0.4146736489398015, 8), (0.41389776479053075, 14), (0.41357014758126004, 13), (0.4133363916796504, 9), (0.4121900166906223, 11), (0.30267493523117395, 1), (0.25626178061170546, 4), (0.25372591683268725, 2), (0.2026368890085653, 5), (0.1915809782296365, 3), (-0.04846902975502297, 7)]
Y_07  :  [(1.0, 7), (-0.03732787909738924, 3), (-0.04846902975502297, 6), (-0.14428361791636357, 1), (-0.17005297451014523, 2), (-0.38855086872647576, 10), (-0.39688833807434615, 13), (-0.3978079544943554, 9), (-0.40022625846607796, 14), (-0.40024840692754143, 12), (-0.40051790906114826, 8), (-0.556018724032774, 4), (-0.601411890692791, 5), (-0.6045267842698097, 11)]
Y_08  :  [(1.0, 8), (0.9678611606648864, 12), (0.9676082830454282, 13), (0.9673905340216551, 14), (0.9665501522705753, 9), (0.845134530530973, 11), (0.8165715824049286, 10), (0.5399477570597936, 4), (0.5397582616820021, 5), (0.4146736489398015, 6), (0.25749435938619814, 2), (0.2516040629755018, 1), (0.09505592667623942, 3), (-0.40051790906114826, 7)]
Y_09  :  [(1.0, 9), (0.968013892228689, 14), (0.9678930990449308, 12), (0.9676368172617493, 13), (0.9665501522705753, 8), (0.8431764236463525, 11), (0.8066274219360859, 10), (0.5350426077994276, 5), (0.533946795654652, 4), (0.4133363916796504, 6), (0.25299286812530203, 2), (0.2461552146334526, 1), (0.09168241279750015, 3), (-0.3978079544943554, 7)]
Y_10  :  [(1.0, 10), (0.8165715824049286, 8), (0.8125600992376579, 12), (0.8113232607418239, 11), (0.8111875603727442, 14), (0.8108378155710994, 13), (0.8066274219360859, 9), (0.6981263213927064, 6), (0.5419338116516438, 4), (0.5299272998756622, 5), (0.3929449218000198, 1), (0.36292380102783184, 2), (0.24675491776253486, 3), (-0.38855086872647576, 7)]
Y_11  :  [(1.0, 11), (0.8459243499774056, 14), (0.845134530530973, 8), (0.8450663763452501, 12), (0.8431764236463525, 9), (0.8427839251690098, 13), (0.8113232607418239, 10), (0.6887736652003423, 5), (0.6206517173616585, 4), (0.4121900166906223, 6), (0.3317785192505406, 1), (0.30691324800315334, 2), (0.15880599536688572, 3), (-0.6045267842698097, 7)]
Y_12  :  [(1.0, 12), (0.9790709954592893, 13), (0.968604311489806, 14), (0.9678930990449308, 9), (0.9678611606648864, 8), (0.8450663763452501, 11), (0.8125600992376579, 10), (0.539042778266852, 4), (0.5386451410104928, 5), (0.41491013123121523, 6), (0.25514608687807955, 2), (0.24813061814812823, 1), (0.09374294416980566, 3), (-0.40024840692754143, 7)]
Y_13  :  [(1.0, 13), (0.9790709954592893, 12), (0.9689465862803777, 14), (0.9676368172617493, 9), (0.9676082830454282, 8), (0.8427839251690098, 11), (0.8108378155710994, 10), (0.5371589859920781, 4), (0.5366804810401934, 5), (0.41357014758126004, 6), (0.2566803013931644, 2), (0.2486376120681256, 1), (0.09537062250924866, 3), (-0.39688833807434615, 7)]
Y_14  :  [(1.0, 14), (0.9689465862803777, 13), (0.968604311489806, 12), (0.968013892228689, 9), (0.9673905340216551, 8), (0.8459243499774056, 11), (0.8111875603727442, 10), (0.542681269994975, 5), (0.5404804991614621, 4), (0.41389776479053075, 6), (0.25729244159981884, 2), (0.2508361285323926, 1), (0.09668211540353398, 3), (-0.40022625846607796, 7)]
```

```py
for seed in range(10):
    chain =  RegressorChain(base_estimator=LR, order='random', random_state=seed).fit(train_x, train_y)
    print(f'seed : {seed} | order : {chain.order}')

    valid_preds = chain.predict(valid_x)
    print(lg_nrmse(valid_y.values, valid_preds))
```
```
seed : 0 | order : random
(1.9537885754103117, [0.2562918577362499, 0.3539190180041992, 0.34996067054150826, 0.19090219863324262, 0.07890862641743011, 0.10345348300996061, 0.1297781457142463, 0.024087184657637675, 0.02397334740033406, 0.03887899777185426, 0.033778074510761244, 0.024132516158132722, 0.024088171380501185, 0.024176046531358537])
```
- seed 1부터 9까지의 결과도 비슷하므로 생략합니다.
